File: src/db/schema.ts
/**
 * Kysely Database interface + domain types.
 * Schema v16 — simplified from v15:
 *   - Merged canvas_grades into canvas_submissions (posted_grade + grade_updated_at)
 *   - Merged canvas_students into students (sortable_name, login_id, sis_user_id, sis_section_id)
 *   - Replaced shadow tables with inline _synced_* columns
 */

import type { Generated, Insertable, Selectable, Updateable } from "kysely";

// ─── Table row types ────────────────────────────────────────────────

export interface MetaTable {
  key: string;
  value: string;
}

export interface StudentsTable {
  canvas_id: number;
  github_username: string | null;
  name: Generated<string>;
  sortable_name: Generated<string>;
  email: Generated<string>;
  login_id: Generated<string>;
  sis_user_id: Generated<string>;
  sis_section_id: Generated<string>;
  excluded: Generated<number>;
}

export interface AssignmentsTable {
  slug: string;
  title: string;
  gh_assignment_slug: string | null;
  canvas_assignment_id: number | null;
  points_possible: Generated<number>;
  deadline: string | null;
}

export interface CanvasAssignmentsTable {
  canvas_id: number;
  name: string;
  points_possible: Generated<number>;
  due_at: string | null;
  published: Generated<number>;
  assignment_group: Generated<string>;
  post_manually: Generated<number>;
  // Inline synced columns (baseline from last pull)
  _synced_name: Generated<string>;
  _synced_points_possible: Generated<number>;
  _synced_due_at: Generated<string | null>;
  _synced_published: Generated<number>;
}

export interface CanvasSubmissionsTable {
  canvas_user_id: number;
  canvas_assignment_id: number;
  submitted: Generated<number>;
  submitted_at: string | null;
  late: Generated<number>;
  lateness_seconds: Generated<number>;
  score: number | null;
  workflow_state: Generated<string>;
  fetched_at: number;
  // Grade fields (merged from canvas_grades)
  posted_grade: Generated<string>;
  grade_updated_at: Generated<number>;
  // Inline synced column (baseline from last pull)
  _synced_posted_grade: Generated<string>;
}

export interface GHStudentsTable {
  github_username: string;
  github_id: Generated<number>;
  name: Generated<string>;
  email: Generated<string>;
  excluded: Generated<number>;
}

export interface GHAssignmentsTable {
  slug: string;
  gh_id: number;
  title: string;
  deadline: string | null;
  points_possible: Generated<number>;
  accepted: Generated<number>;
  submissions_count: Generated<number>;
  passing_count: Generated<number>;
  starter_code_repo: Generated<string>;
  submittable_files: Generated<string>;
}

export interface GHSubmissionsTable {
  github_username: string;
  assignment_slug: string;
  submitted: Generated<number>;
  late: Generated<number>;
  lateness_seconds: Generated<number>;
  repo_name: Generated<string>;
  commits_after_deadline: Generated<number>;
  commit_count: Generated<number>;
  passing: Generated<number>;
  gh_autograder_score: Generated<string>;
  last_commit_at: Generated<string>;
  last_commit_sha: Generated<string>;
  fetched_at: number;
}

// ─── Database interface ─────────────────────────────────────────────

export interface Database {
  meta: MetaTable;
  students: StudentsTable;
  assignments: AssignmentsTable;
  canvas_assignments: CanvasAssignmentsTable;
  canvas_submissions: CanvasSubmissionsTable;
  gh_students: GHStudentsTable;
  gh_assignments: GHAssignmentsTable;
  gh_submissions: GHSubmissionsTable;
}

// ─── Convenience aliases ────────────────────────────────────────────

export type TableName = keyof Database;

export type Meta = Selectable<MetaTable>;
export type Student = Selectable<StudentsTable>;
export type Assignment = Selectable<AssignmentsTable>;
export type CanvasAssignment = Selectable<CanvasAssignmentsTable>;
export type CanvasSubmission = Selectable<CanvasSubmissionsTable>;
export type GHStudent = Selectable<GHStudentsTable>;
export type GHAssignment = Selectable<GHAssignmentsTable>;
export type GHSubmission = Selectable<GHSubmissionsTable>;

export type NewStudent = Insertable<StudentsTable>;
export type NewAssignment = Insertable<AssignmentsTable>;
export type NewCanvasAssignment = Insertable<CanvasAssignmentsTable>;
export type NewCanvasSubmission = Insertable<CanvasSubmissionsTable>;
export type NewGHStudent = Insertable<GHStudentsTable>;
export type NewGHAssignment = Insertable<GHAssignmentsTable>;
export type NewGHSubmission = Insertable<GHSubmissionsTable>;

export type StudentUpdate = Updateable<StudentsTable>;
export type CanvasAssignmentUpdate = Updateable<CanvasAssignmentsTable>;
export type CanvasSubmissionUpdate = Updateable<CanvasSubmissionsTable>;
export type GHStudentUpdate = Updateable<GHStudentsTable>;

File: cass/db/schema.py
"""Domain models for cass.

User-facing types for students, assignments, submissions, grades.
Each model corresponds to either a DB table or a virtual view.
"""

from __future__ import annotations

__docformat__ = "google"

from datetime import datetime
from typing import TYPE_CHECKING

import msgspec
import msgspec.structs

if TYPE_CHECKING:
    from ..apis.canvas.schema import CanvasAssignmentResponse, CanvasStudentResponse
    from ..apis.github.schema import GHAssignmentResponse, GHStudentInfo

# ===================================================================
# User-facing tables (Canvas)
# ===================================================================


class CanvasStudent(msgspec.Struct):
    """A Canvas LMS student record."""

    canvas_id: int
    name: str = ""
    sortable_name: str = ""
    email: str = ""
    login_id: str = ""
    sis_user_id: str = ""
    sis_section_id: str = ""

    @classmethod
    def from_api(
        cls,
        s: CanvasStudentResponse,
        *,
        sis_section_id: str = "",
    ) -> CanvasStudent:
        """Convert a Canvas API student response to a domain model."""
        return cls(
            canvas_id=s.id,
            name=s.name,
            sortable_name=s.sortable_name,
            email=s.email,
            login_id=s.login_id,
            sis_user_id=s.sis_user_id or "",
            sis_section_id=sis_section_id,
        )


class CanvasAssignment(msgspec.Struct):
    """A Canvas LMS assignment."""

    canvas_id: int
    name: str = ""
    points_possible: float = 0.0
    due_at: str | None = None
    published: bool = False
    assignment_group: str = ""
    post_manually: bool = False

    @classmethod
    def from_api(
        cls,
        a: CanvasAssignmentResponse,
        *,
        group_name: str = "",
    ) -> CanvasAssignment:
        """Convert a Canvas API assignment response to a domain model."""
        return cls(
            canvas_id=a.id,
            name=a.name,
            points_possible=a.points_possible or 0.0,
            due_at=a.due_at or None,
            published=a.published,
            assignment_group=group_name,
            post_manually=a.post_manually,
        )


class CanvasSubmission(msgspec.Struct):
    """A Canvas LMS submission record."""

    canvas_user_id: int
    canvas_assignment_id: int
    submitted: bool = False
    submitted_at: datetime | None = None
    late: bool = False
    lateness_seconds: int = 0
    score: float | None = None
    workflow_state: str = ""


class CanvasGrade(msgspec.Struct):
    """A grade ready for Canvas API push."""

    canvas_user_id: int
    canvas_assignment_id: int
    score: float | None = None
    posted_grade: str = ""


# ===================================================================
# User-facing tables (GitHub)
# ===================================================================


class GHStudent(msgspec.Struct):
    """A GitHub Classroom roster student."""

    github_username: str
    github_id: int = 0
    name: str = ""
    email: str = ""
    excluded: bool = False

    @classmethod
    def from_api(cls, s: GHStudentInfo) -> GHStudent:
        """Convert a GHStudentInfo API aggregate to a domain model."""
        return cls(
            github_username=s.login.lower(),
            github_id=int(s.id) if s.id else 0,
            name=s.name,
            email=s.email,
        )


class GHAssignment(msgspec.Struct):
    """A GitHub Classroom assignment."""

    slug: str
    gh_id: int = 0
    title: str = ""
    deadline: str | None = None
    points_possible: float = 1.0
    accepted: int = 0
    submissions_count: int = 0
    passing_count: int = 0
    starter_code_repo: str = ""
    submittable_files: str = ""

    @classmethod
    def from_api(cls, a: GHAssignmentResponse) -> GHAssignment:
        """Convert a GHAssignmentResponse API type to a domain model."""
        starter = (
            a.starter_code_repository.full_name
            if a.starter_code_repository and a.starter_code_repository.full_name
            else ""
        )
        return cls(
            slug=a.slug,
            gh_id=a.id,
            title=a.title,
            deadline=a.deadline or None,
            points_possible=1.0,
            accepted=a.accepted,
            submissions_count=a.submissions,
            passing_count=a.passing,
            starter_code_repo=starter,
        )


class GHSubmission(msgspec.Struct):
    """A GitHub Classroom submission record."""

    github_username: str
    assignment_slug: str
    submitted: bool = False
    late: bool = False
    lateness_seconds: int = 0
    repo_name: str = ""
    commits_after_deadline: int = 0
    commit_count: int = 0
    passing: bool = False
    gh_autograder_score: str = ""
    last_commit_at: str = ""
    last_commit_sha: str = ""

    def with_assignment_slug(self, slug: str) -> GHSubmission:
        """Return a copy with a different assignment_slug."""
        return msgspec.structs.replace(self, assignment_slug=slug)


# ===================================================================
# View models (no backing table)
# ===================================================================


class GHRecentCommit(msgspec.Struct):
    """A row in the enriched GH recent commits view (live query, no table)."""

    github_username: str = ""
    assignment_slug: str = ""
    student: str = ""
    assignment_name: str = ""
    last_commit_at: str = ""
    commit_count: int = 0
    commit_url: str = ""
    repo_url: str = ""
    late: bool = False
    last_commit_sha: str = ""


# ===================================================================
# Internal/master tables
# ===================================================================


class Student(msgspec.Struct):
    """A student in the unified course roster (Canvas-authoritative)."""

    canvas_id: int
    github_username: str = ""
    name: str = ""
    email: str = ""
    excluded: bool = False

    @property
    def handle_lower(self) -> str:
        return self.github_username.lower()

    @property
    def display_name(self) -> str:
        return self.name or self.github_username or str(self.canvas_id)


class Assignment(msgspec.Struct):
    """A unified assignment mapping GH Classroom and Canvas."""

    slug: str
    title: str
    gh_assignment_slug: str = ""
    canvas_assignment_id: int = 0
    points_possible: float = 0.0
    deadline: datetime | None = None
